SSGAC changes

seanclee opened the pull request:
Here are brief descriptions of our changes:

## Major Changes
- **Pandas dataframes can now be passed directly to LDSC with direct calls from Python.**
  * _Implemented in script `ldscore.sumstats.py` lines [353-363, 453-463]_
  * While the original `bulik/ldsc` is callable from command line, we attempted to make it more flexible so that functions can be called directly in Python. Specifically, we modified the `estimate_h2` and `estimate_rg` functions which now “recognize” the running environment. If the summary statistics are already provided in the format of pandas dataframes (in Python), `ldsc` will input them directly without going through the import & read steps. The parsing functionality is retained but is reserved for calling LDSC via command line.  
- **--rg now calculates pairwise genetic correlations for all traits.**
  * _Implemented in script `ldscore.sumstats.py` lines [478-509] (for-loops)_
  * The current `bulik/ldsc` can run multiple bivariate LDSC at the same time, but it only estimates the genetic correlations between _one_ trait and a list of other traits. We extended this function to allow for pairwise estimation across _all_ traits. The results will be stored in a matrix object, which is returned by `estimate_rg`, along with a list of tables containing rg-related statistics. More updates to the output options are discussed below. 
- **`--h2` can now run univariate LDSC on a list of traits sequentially.**
  * _Implemented in script `ldscore.sumstats.py` lines [368-412] (for loops)_
  * The updated `ldsc` package can take in multiple summary statistics using a single command and estimate the heritability for each of the traits sequentially. The results are conveniently stored in a list of tables containing h2-related statistics. Users may choose to view (and operate on) these estimates directly in Python or write them to files. 
- **Estimate results from `h2` and `rg` can be printed to .txt or excel tables.**
  * _Implemented in script `ldscore.sumstats.py` functions: `write_h2_table`, `write_rg_table`, `print_estimates`, `print_cov` and `print_delete_vals`. Also reflected in script `ldsc.py`, new flags `--write-excel` and `--print-estimates`_
  * We added four output options (argument group “Output Options”) so that the estimate results can be presented based on users’ preferences.
- **`ldsc` can now exclude a list of SNPs when running LDSC regressions.**
  * _Implemented in script `ldscore.sumstats.py` lines [265-273]. Also reflected in script `ldsc.py`, new flag `--exclude`_
  * This new option may be useful when performing sensitivity analysis that requires excluding SNPs from a high effect locus without having to generate a new set of summary statistics first. 
- **The LD reference data is read in only once, which can lead to substantial improvement in runtime when analyzing multiple traits.** 
  * _Implemented in script `ldscore.sumstats.py` functions `_read_ref`, `_read_ld_sumstats`, and places wherever they are called_
  * We removed redundant readings of the reference panels, by importing the LD score data only once throughout the analysis. This significantly saves the `ldsc` runtime without influencing the estimate results.
- **`--two-step ` is set to infinity by default.**
  * _Implemented in script `ldscore.sumstats.py` lines [389-390, 490-491]_
  * The current package by default imposes a cutoff value of 30 on the test statistic in two-step estimation. We removed this limit by setting its default value to infinity. As a result, SNPs with high chi^2 values are no longer filtered out.
- **Useful objects related to allele relationships are stored in a separate module `allele_info.py`**
  * _Implemented in script `allele_info.py`_

## Minor Changes
- **`--ref-ld` will use the reference panel from `--w-ld` if it is not specified (and vice versa).**
  * _Implemented in script `ldsc.py` lines [651-665]_
  * Only one of `--ref-ld` and `--w-ld` is required, unless the flag `--overlap-annot` is specified.
- **The logging module is used to create log files and stream output to console.**
  * _Implemented across scripts_
- **The argument options in `ldsc.py` are now organized into groups by functionality.**
  * _Implemented in script `ldsc.py` lines [409-595]_

Please direct any questions/comments to Omeed Maghzian (maghzian@nber.org), Hui Li (huilieatpraylove@gmail.com), and Sean Lee (seanclee@nber.org).

tpoterba commented on ldscore/sumstats.py:
```diff
@@ -309,78 +332,136 @@ def cell_type_specific(args, log):
+        h2_list = h2_paths # columns
+        n_pheno = len(h2_paths)
+    else: # python case
+        h2_frames = args.h2 # rows
```
assert the type is a dataframe here

tpoterba commented on ldscore/sumstats.py:
```diff
@@ -309,78 +332,136 @@ def cell_type_specific(args, log):
+    H2 = []
+    sumstats_list = {}
+
+    if type(args.h2) == str: # command line case
```
`if isinstance(args.h2, str)` is more pythonic

tpoterba commented on ldscore/sumstats.py:
```diff
@@ -309,78 +332,136 @@ def cell_type_specific(args, log):
+
+        ref_ld, w_ld, M_annot, ref_ld_cnames, novar_cols=map(lambda x: x.copy(),(ref_ld_orig, w_ld_orig, M_annot_orig, ref_ld_cnames_orig, novar_cols_orig))
+
+        if type(args.h2) == str:
```
see above

tpoterba commented on ldscore/sumstats.py:
```diff
@@ -309,78 +332,136 @@ def cell_type_specific(args, log):
+    RG = []
+    sumstats_list = {}
+
+    if type(args.rg) == str: # command line case
```
see above

tpoterba commented on munge_sumstats.py:
```diff
@@ -440,111 +443,129 @@ def allele_merge(dat, alleles, log):
+                    help="Use this flag to parse Stephan Ripke's daner* file format.")
+filters.add_argument('--daner-n', default=False, action='store_true',
+                    help="Use this flag to parse more recent daner* formatted files, which "
+            "include sample size column 'Nca' and 'Nco'.")
```
indentation

tpoterba commented on munge_sumstats.py:
```diff
@@ -440,111 +443,129 @@ def allele_merge(dat, alleles, log):
+filters.add_argument('--chunksize', default=5e6, type=int,
+                    help='Chunksize.')
+parser.add_argument('--keep-str-ambig', default=False, action='store_true',
+                    help=argparse.SUPPRESS) # This options allows munge sumstats to retain strand ambiguous SNPS instead of dropping them.
```
why not put this message in `help`?

seanclee replied:
Hi @tpoterba,

We added this flag so that it doesn't remove strand ambiguous alleles (TA, GC) by default. However, Raymond suggested that for the purposes of generic munging, this option would be better if kept invisible.

tpoterba commented:
another question - why the `xrange` to `range` changes?

seanclee commented:
@tpoterba Sorry for the confusion. I tried to update my forked repo with the newly merged PR on ldsc but doing so complicated the commit history, so I just reverted the changes.

To answer your question about changing `xrange` to `range`, I updated the ldsc such that it is compatible with python 3. Since `xrange` is not existent in python 3, I changed all such instances to `range`.

seanclee commented:
After careful consideration, we decided not to push out changes to the main repo. Thanks!
